nx/lib/nx/defn/kernel.ex

```diff
@@ -1013,4 +1013,129 @@ defmodule Nx.Defn.Kernel do
   defmacro @expr do
     quote do: Kernel.@(unquote(expr))
   end
+
+  @doc """
+  Asserts the `tensor` has a certain `shape`.
+
+  If it succeeds, it returns the given tensor. Raises
+  an error otherwise.
+
+  ## Examples
+
+  To assert the tensor is a scalar, you can pass the empty tuple `shape`:
+
+      iex> assert_shape Nx.tensor(13), {}
+      #Nx.Tensor<
+        s64
+        13
+      >
+
+  If the shapes do not match, an error is raised:
+
+      iex> assert_shape Nx.tensor([1, 2, 3]), {}
+      ** (ArgumentError) expected tensor to be a scalar, got tensor with shape {3}
+
+      iex> assert_shape Nx.tensor([1, 2, 3]), {4}
+      ** (ArgumentError) expected tensor to have shape {4}, got tensor with shape {3}
+
+  If you want to assert on the rank or shape patterns, use
+  `assert_shape_pattern/2` instead.
+  """
+  def assert_shape(tensor, shape) when is_tuple(shape) do
+    case Nx.shape(tensor) do
+      ^shape ->
+        tensor
+
+      other ->
+        raise ArgumentError,
+              "expected tensor to #{shape_to_string(shape)}, got tensor with shape #{inspect(other)}"
+    end
+  end
+
+  defp shape_to_string({}), do: "be a scalar"
+  defp shape_to_string(shape), do: "have shape " <> inspect(shape)
+
+  @doc """
+  Asserts the `tensor` has a certain `shape` pattern.
+
+  If it succeeds, it returns the given tensor. Raises
+  an error otherwise.
+
+  ## Examples
+
+  Opposite to `assert_shape/2`, where the given shape is a value,
+  `assert_shape_pattern` allows the shape to be any Elixir pattern.
+  We can use this to match on ranks:
+
+      iex> assert_shape_pattern Nx.tensor([[1, 2], [3, 4]]), {_, _}
+      #Nx.Tensor<
+        s64[2][2]
+        [
+          [1, 2],
+          [3, 4]
+        ]
+      >
+
+      iex> assert_shape_pattern Nx.tensor([1, 2, 3]), {_, _}
+      ** (ArgumentError) expected tensor to match shape {_, _}, got tensor with shape {3}
+
+  Or even use variables to assert on properties such as square matrices:
+
+      iex> assert_shape_pattern Nx.tensor([[1, 2], [3, 4]]), {x, x}
+      #Nx.Tensor<
+        s64[2][2]
+        [
+          [1, 2],
+          [3, 4]
+        ]
+      >
+
+      iex> assert_shape_pattern Nx.tensor([1, 2, 3]), {x, x}
+      ** (ArgumentError) expected tensor to match shape {x, x}, got tensor with shape {3}
+
+  You can also use guards to specify tall matrices and so forth:
+
+      iex> assert_shape_pattern Nx.tensor([[1], [2]]), {x, y} when x > y
+      #Nx.Tensor<
+        s64[2][1]
+        [
+          [1],
+          [2]
+        ]
+      >
+
+      iex> assert_shape_pattern Nx.tensor([1, 2]), {x, y} when x > y
+      ** (ArgumentError) expected tensor to match shape {x, y} when x > y, got tensor with shape {2}
+
+  """
+  defmacro assert_shape_pattern(tensor, shape) do
```

Comment:
Is it possible to extend this such that we can support multiple patterns? e.g. a tensor can be `{_, _} or {_, _, _}`, etc.

Reply:
Or perhaps we can support a wildcard pattern for arbitrary ranks, e.g. for `conv` the shape of the `input` must be:

```elixir
{batch, features, ...any spatial ranks...}
```

And the kernel needs:

```elixir
{out_features, features, ...kernel size...}
```

Reply:
@seanmor5 you would be able to do this:

    assert_shape_pattern tensor, shape when tuple_size(shape) in [2, 3]

But I w7ld say at this point we are way too much into the weeds. Probably `assert_rank` is necessary in those cases?

Reply:
I think what @seanmor5 is getting into is that sometimes we want to assert that the shape ends in a certain shape, like when we work with tensor batches.

In these cases a simple assertion on the rank wouldn't suffice. For lists this is easy through `[a, b | _]` matching, though I'm not sure if we can (or want to) have this kind of ellipsis matching

Reply:
We could introduce a new notation, such as `{..., a, b}` or something of sorts, but then we are straying a bit from Elixir. For now, you can do something like `tuple when elem(tuple, 0) == batch and elem(tuple, 1) == features` but I don't think that's much better?

Reply:
Yeah... Perhaps we could have assert_shape also accept a list, which would allow for at least cons matching.
Not sure about that, though.

I do think that the best way for more specific cases is using elem.

Perhaps allowing a %{index => axis_size} map to be passed could be a compromise. 

For maps it would be a sequence of assert elem(shape, index) == axis_size

Reply:
The list doesn't allow suffix match, so it would help in some cases but not others. If we are going to add special semantics, then I think `...` is probably the route I would go. The big question for now is if we think this function is useful as is or if we want to wait until we can match on those more complex patterns?

Reply:
I have done more thinking about this and I think the need for these kind of special assertions becomes less necessary with named Tensors because we already have rules for how dimensions should interact based on matching names. 

So I think we is the current implementation is fine :)